src/serializers/filter.rs

```diff
@@ -14,6 +15,48 @@ pub(crate) struct SchemaFilter<T> {
     exclude: Option<AHashSet<T>>,
 }
 
+fn map_negative_index<'a, 'py: 'a>(py: Python<'py>, value: &'a PyAny, len: Option<usize>) -> PyResult<&'a PyAny> {
+    match len {
+        Some(len) => Ok(value
+            .call_method1(intern!(py, "__mod__"), (len,))
+            .map_or_else(|_| value, |v| v)),
+        None => {
+            // check that it's not negative
+            let negative: bool = value.call_method1(intern!(py, "__lt__"), (0,))?.extract()?;
+            if negative {
+                Err(PyValueError::new_err(
+                    "Negative indices cannot be used to exclude items on unsized iterables",
+                ))
+            } else {
+                Ok(value)
+            }
+        }
+    }
+}
+
+fn map_negative_indices<'a, 'py: 'a>(
+    py: Python<'py>,
+    include_or_exclude: &'a PyAny,
+    len: Option<usize>,
+) -> PyResult<&'a PyAny> {
+    if let Ok(exclude_dict) = include_or_exclude.downcast::<PyDict>() {
+        let out = PyDict::new(py);
+        for (k, v) in exclude_dict.iter() {
+            out.set_item(map_negative_index(py, k, len)?, v)?;
+        }
+        Ok(out)
+    } else if let Ok(exclude_set) = include_or_exclude.downcast::<PySet>() {
+        let mut values = Vec::with_capacity(exclude_set.len());
+        for v in exclude_set {
+            values.push(map_negative_index(py, v, len)?)
+        }
+        Ok(PySet::new(py, values)?)
+    } else {
+        // return as is and deal with the error later
+        Ok(include_or_exclude)
+    }
+}
```

Comment:
This is not particularly efficient. We're downcasting twice (once here, again later) and re-creating the entire iterable (when really we could do it on the fly). One option to change this would be to duplicate the "hellish logic" below (which I wanted to avoid) to handle indexes and keys/fields separately. It might also be good to extract the entirety of `include` and `exclude` into rust types upfront so that it's not only cheaper to do this post-processing but also maybe some of the "hellish logic" becomes simpler or at least more understandable via type hints.

A quick win might be to check if there are negative indexes in a first pass and only map them / copy the input if there are. I'm skipping that for now since it requires benchmarks to prove it's even better and is a lot more code/boilerplate.

I think this is good enough (and fast enough) for now, we can come back to those more advanced options later.